src/noise/handshake.rs

```diff
@@ -418,44 +427,26 @@ impl Handshake {
         &mut self,
         packet: HandshakeResponse,
     ) -> Result<Session, WireGuardError> {
-        let state = std::mem::replace(&mut self.state, HandshakeState::None);
-
-        let (mut chaining_key, mut hash, ephemeral_private, local_index, time_sent) = match state {
-            HandshakeState::InitSent {
-                chaining_key,
-                hash,
-                ephemeral_private,
-                local_index,
-                time_sent,
-                ..
-            } => (
-                chaining_key,
-                hash,
-                ephemeral_private,
-                local_index,
-                time_sent,
-            ),
-            _ => {
-                std::mem::replace(&mut self.state, state);
-                return Err(WireGuardError::UnexpectedPacket);
-            }
+        // Check if there is a handshake awaiting a response and return the correct one
+        let (state, is_previous) = match (&self.state, &self.previous) {
+            (HandshakeState::InitSent(s), _) if s.local_index == packet.receiver_idx => (s, false),
+            (_, HandshakeState::InitSent(s)) if s.local_index == packet.receiver_idx => (s, true),
+            _ => return Err(WireGuardError::UnexpectedPacket),
```

Comment:
I presume it's intentional that state is no longer reset on error.

Reply:
Yes